### PR TITLE
[Core] Do not report stats when worker is already dead.

### PR DESCRIPTION
**Pull request by rkooo567**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -3306,6 +3306,9 @@ void NodeManager::HandleGetNodeStats(const rpc::GetNodeStatsRequest &node_stats_
     return;
   }
   for (const auto &worker : all_workers) {
+    if (worker->IsDead()) {
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:3309]: Does it cause line3332(send_reply_callback(Status::OK(), nullptr, nullptr);) not to be called?

Comment (ffbin) [reply src/ray/raylet/node_manager.cc:3309]: I prefer to add a default parameter to `GetAllRegisteredWorkers` and `GetAllRegisteredDrivers` to filter the worker and driver of alive (not filtered by default).

Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:3309]: Sounds good. Let me fix this.

Comment (edoakes) [reply src/ray/raylet/node_manager.cc:3309]: Shouldn't dead workers be filtered from those calls anyways? When do they actually get removed?

Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:3309]: It doesn't look like it. It only returns "Registered" workers, and there seem to be cases where there are dead registered workers. Maybe when the actor is marked as dead and deregistered are different. After I introduced this line, I saw some core worker stats failures logs from raylet have disappeared.

**Comment by rkooo567** [comment]
Addressed comments @edoakes @ffbin 

For other functions, I feel like they could also set `filter_dead-*` to be true, but I just preserved the previous behavior just in case.

File: src/ray/raylet/node_manager.cc [hunk 2]
```diff
@@ -496,8 +496,10 @@ void NodeManager::Heartbeat() {
 }
 
 void NodeManager::DoLocalGC() {
-  auto all_workers = worker_pool_.GetAllRegisteredWorkers();
-  for (const auto &driver : worker_pool_.GetAllRegisteredDrivers()) {
+  auto all_workers =
+      worker_pool_.GetAllRegisteredWorkers(false /* filter_dead_workers */);
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:500]: How about set a default value of filter_dead_workers? The previous implementation default is false.

Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:500]: Ah yeah. I tried, and it seemed to have a compiler error. But I think I just did it wrong. Is this the right way to do that? 

https://www.programiz.com/cpp-programming/default-argument

Comment (ffbin) [reply src/ray/raylet/node_manager.cc:500]: Yeah, this is the right way.

Comment (edoakes) [inline src/ray/raylet/node_manager.cc:500]:
```suggestion
      worker_pool_.GetAllRegisteredWorkers(/* filter_dead_workers=*/false);
```
to stay consistent with other places in the code. same with below

**Review by edoakes — APPROVED** [review]

**Review by ffbin — APPROVED** [review]

**Pull request merged**